task/sp_measure/src/main.rs

```diff
@@ -0,0 +1,114 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#![no_std]
+#![no_main]
+
+use drv_sp_ctrl_api::*;
+use ringbuf::*;
+use sha2::{Digest, Sha256};
+use userlib::*;
+
+const READ_SIZE: usize = 256;
+
+const TRANSACTION_SIZE: u32 = 1024;
+
+task_slot!(SP_CTRL, swd);
+
+#[derive(Copy, Clone, PartialEq)]
+enum Trace {
+    ErrCnt(usize),
+    Addr(u32),
+    Start(u64),
+    End(u64),
+    Data([u8; READ_SIZE]),
+    ShaGood,
+    ShaBad,
+    None,
+}
+
+ringbuf!(Trace, 16, Trace::None);
+
+#[export_name = "main"]
+fn main() -> ! {
+    let mut err_cnt = 0;
+    loop {
+        let mut sha = Sha256::new();
+        let sp_ctrl = SpCtrl::from(SP_CTRL.get_task_id());
+
+        match sp_ctrl.setup() {
+            Err(_) => panic!(),
+            _ => (),
+        }
+
+        let mut data: [u8; READ_SIZE] = [0; READ_SIZE];
+
+        let start = sys_get_timer().now;
+        ringbuf_entry!(Trace::Start(start));
+        for addr in (FLASH_START..FLASH_END).step_by(READ_SIZE) {
+            if addr % TRANSACTION_SIZE == 0 {
+                loop {
+                    match sp_ctrl
+                        .read_transaction_start(addr, addr + TRANSACTION_SIZE)
+                    {
+                        Err(_) => {
+                            err_cnt += 1;
+                            let _ = sp_ctrl.setup();
+                            continue;
+                        }
+                        _ => break,
+                    }
+                }
+            }
+
+            data.fill(0);
+            loop {
+                match sp_ctrl.read_transaction(&mut data) {
+                    Err(_) => {
+                        ringbuf_entry!(Trace::Addr(addr));
+                        ringbuf_entry!(Trace::Data(data));
+                        loop {
+                            match sp_ctrl.setup() {
+                                Err(_) => continue,
+                                Ok(_) => {
+                                    err_cnt += 1;
+                                    match sp_ctrl.read_transaction_start(
+                                        addr,
+                                        addr + TRANSACTION_SIZE,
+                                    ) {
+                                        Err(_) => continue,
+                                        Ok(_) => break,
+                                    }
+                                }
+                            }
+                        }
+                    }
+                    Ok(_) => break,
+                }
+            }
+
+            sha.update(&data);
```

Comment:
At some point, should we use the hardware HASH peripheral, or does SWD overhead eat most of our time?

Reply:
I'm about to go update this data in RFD32. Initially we had planned on using sha512 so that we could share code with the salty library. The hash function built into salty doesn't implement any of the RustCrypto hash traits so it's not usable in an hmac. I've since switched to using sha3-256 but I never went back to the RFD to fix this.